### PR TITLE
Poisson Cost Function

### LINKED ISSUE #14490: Add Poisson cost function
The correct cost function formula for Poisson statistics should be (for bin contents=N, fitted mean count rate=y)
y<0: forbidden (infinite cost)
y>=0, N=0: cost function=2_y
y>=0, N>0: cost function=2_( (y-N) + N*(log(N)-log(y)) )

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
This finishes Roman's work on Poisson cost functions (started in #25686)

@alicerussell1 could you add release notes if required, as you are more familiar with the potential benefits of this technique.

Changes since the original PR:
- Squashes the original commit series and resolve conflicts with master
- Added unit tests for the methods
- Moved protected methods into public following the conventions of the base classes
- Fixed a bug where using any parameters would result in an OOB
- Fixed a bug where the determinant was always calculated instead of respecting the flag

**To test:**
- Ensure unit tests pass

```
Load some muon data and plot a few spectra
Fit to the spectra using least squares and poisson cost functions
The parameters should be similar but different.
```

Fixes #14490 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by alicerussell1 — COMMENTED** [review]
This seems to work as it should and the added tests look good. I've added a few comments below which are tiny things. 

I reckon we should include some release notes but it probably doesn't need to be anything more than a new cost function poisson has been added for use in fitting. Documentation with more of the specific details will be added but that can be done in a separate PR - there is already an issue for it #26732.

File: Framework/CurveFitting/src/CostFunctions/CostFuncFitting.cpp [hunk 1]
```diff
@@ -278,6 +280,232 @@ void CostFuncFitting::getParameters(GSLVector &params) const {
...
+  }
+
+  auto seqDomain = boost::dynamic_pointer_cast<SeqDomain>(m_domain);
+
+  if (seqDomain) {
+    seqDomain->additiveCostFunctionValDerivHessian(*this, evalDeriv,
+                                                   evalHessian);
+  } else {
+    if (!m_values) {
+      throw std::runtime_error("LeastSquares: undefined FunctionValues.");
```
Comment (alicerussell1) [inline Framework/CurveFitting/src/CostFunctions/CostFuncFitting.cpp]:
```suggestion
      throw std::runtime_error("CostFunction: undefined FunctionValues.");
```

File: Framework/CurveFitting/src/CostFunctions/CostFuncFitting.cpp [hunk 2]
```diff
@@ -278,6 +280,232 @@ void CostFuncFitting::getParameters(GSLVector &params) const {
...
+  }
+  return m_hessian;
+}
+
+/**
+ * Save current parameters, derivatives and hessian.
+ */
+void CostFuncFitting::push() {
+  if (m_pushed) {
+    throw std::runtime_error("Least squares: double push.");
```
Comment (alicerussell1) [inline Framework/CurveFitting/src/CostFunctions/CostFuncFitting.cpp]:
```suggestion
    throw std::runtime_error("CostFunction: double push.");
```

File: Framework/CurveFitting/src/CostFunctions/CostFuncFitting.cpp [hunk 3]
```diff
@@ -278,6 +280,232 @@ void CostFuncFitting::getParameters(GSLVector &params) const {
...
+  getParameters(m_pushedParams);
+  m_pushed = true;
+}
+
+/**
+ * Restore saved parameters, derivatives and hessian.
+ */
+void CostFuncFitting::pop() {
+  if (!m_pushed) {
+    throw std::runtime_error("Least squares: empty stack.");
```
Comment (alicerussell1) [inline Framework/CurveFitting/src/CostFunctions/CostFuncFitting.cpp]:
```suggestion
    throw std::runtime_error("CostFunction: empty stack.");
```

**Review by alicerussell1 — CHANGES_REQUESTED** [review]

**Review by alicerussell1 — APPROVED** [review]

File: Framework/CurveFitting/src/Algorithms/CalculateChiSquared.cpp [hunk 4]
```diff
@@ -250,7 +250,7 @@ class ChiSlice {
                               bool &ok) {
 
     auto base = ChebfunBase::bestFitAnyTolerance(lBound, rBound, *this, P, A,
-                                                 1.0, 1e-4, 129);
+                                                 1e-4, 129, 1.0);
```
Comment (martyngigg) [inline Framework/CurveFitting/src/Algorithms/CalculateChiSquared.cpp]: Could you help me understand why this change is required?

The definition of `bestFitAnyTolerance` says that the last 3 parameters are `maxA`, `tolerance` & `maxSize` so it would imply that the tolerance has changed to `129`? Is that correct?

File: Framework/CurveFitting/src/GSLMatrix.cpp [hunk 5]
```diff
@@ -101,8 +101,16 @@ bool GSLMatrix::isEmpty() const { return m_data.empty(); }
 /// @param nx :: New first dimension
 /// @param ny :: New second dimension
 void GSLMatrix::resize(const size_t nx, const size_t ny) {
-  m_data.resize(nx * ny);
-  m_view = gsl_matrix_view_array(m_data.data(), nx, ny);
+  if (nx == 0 && ny == 0) {
```
Comment (martyngigg) [inline Framework/CurveFitting/src/GSLMatrix.cpp]: Is there an issue here if either `nx` or `ny` are `0`?

**Review by martyngigg — COMMENTED** [review]
Just a couple of additional questions about 2 places.

File: Framework/CurveFitting/src/Algorithms/CalculateChiSquared.cpp (hunk 4 above)
Comment (DavidFair) [reply Framework/CurveFitting/src/Algorithms/CalculateChiSquared.cpp]: I'd have to defer to @alicerussell1 , as this is a section of Roman's PR that I've inherited but not touched

File: Framework/CurveFitting/src/GSLMatrix.cpp (hunk 5 above)
Comment (DavidFair) [reply Framework/CurveFitting/src/GSLMatrix.cpp]: Yes, GSL 2.3 (RHEL 7) will not accept a matrix / vector where any n = 0
In GSL >=2.4 this was relaxed so it will work

I had a look how existing code in the class dealt with this, and it appears that we always set a length of 1 and filled with zeros in this edge case.
I'm not sure why a user would have no parameters but it's something which previously threw an abort so I made our side deal with it as sanely as possible

File: Framework/CurveFitting/src/Algorithms/CalculateChiSquared.cpp (hunk 4 above)
Comment (alicerussell1) [reply Framework/CurveFitting/src/Algorithms/CalculateChiSquared.cpp]: I can't see a reason for this, I think it might be a mistake that should probably be changed back

Comment (martyngigg) [reply Framework/CurveFitting/src/Algorithms/CalculateChiSquared.cpp]: Fixed in #26815
